### PR TITLE
(MODULES-1960) Invoke-DscResource

### PR DESCRIPTION
**Pull request by jpogran**
~~Depends on #11 to be pulled in first.~~

### CODE REVIEW

**Comment by Iristyle** [comment]
I believe we're going to want some basic tests like those in place that use `ps_script_content`.

See https://github.com/puppetlabs/puppetlabs-dsc/blob/master/build/dsc/templates/dsc_type_spec.rb.erb#L163-L238

File: lib/puppet_x/msutter/helpers/dsc_provider_helpers.rb [hunk 1]
```diff
@@ -61,6 +61,8 @@ def native_path(path)
           path.gsub(File::SEPARATOR, File::ALT_SEPARATOR)
         end
 
+        def dsc_invoke_method
```
Comment (Iristyle) [inline lib/puppet_x/msutter/helpers/dsc_provider_helpers.rb]: Do we need this defined in the helper, or can this be local state set in the creation of the templated Powershell?

File: lib/puppet/provider/base_dsc/powershell.rb [hunk 2]
```diff
@@ -60,5 +60,10 @@ def ps_script_content(mode)
     template.result(binding)
   end
 
-
+  def ps_invoke_content(mode)
```
Comment (Iristyle) [inline lib/puppet/provider/base_dsc/powershell.rb]: @jpogran I believe you mentioned that we could just reuse `ps_script_content` given we're fully on board with using `Invoke-DscResource` now... but it looks like `ps_invoke_content` is still kicking around.

File: lib/puppet_x/msutter/templates/invoke_dsc_resource.erb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+$result = Invoke-DscResource @invokeParams
+
+switch($Method){
+  'Test'{
+    # true if what is expected
+    # false if it is not what is expected
+    $returnValue = [bool]($result.InDesiredState)
+  }
+  'Set'{
+    if($result.ReturnValue -eq $Null){
```
Comment (Iristyle) [inline lib/puppet_x/msutter/templates/invoke_dsc_resource.erb]: Simplify?

```powershell
$returnValue = @($Null, 0) -contains $result.ReturnValue
```

File: lib/puppet_x/msutter/templates/invoke_dsc_resource.erb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+<% end -%>
+}
+
+$result = Invoke-DscResource @invokeParams
+
+switch($Method){
+  'Test'{
+    # true if what is expected
+    # false if it is not what is expected
+    $returnValue = [bool]($result.InDesiredState)
```
Comment (Iristyle) [inline lib/puppet_x/msutter/templates/invoke_dsc_resource.erb]: This still seems a little odd - did we get to the bottom of what this value is supposed to be "officially"?

File: lib/puppet_x/msutter/templates/invoke_dsc_resource.erb [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    }elseif($result.ReturnValue -eq 0){
+      $returnValue = $true
+    }else{
+      $returnValue = $false
+    }
+  }
+}
+
+if($currentRefreshMode){
+  Set-RefreshMode $currentRefreshMode
```
Comment (Iristyle) [inline lib/puppet_x/msutter/templates/invoke_dsc_resource.erb]: If anything fails above, then we won't reset the LCM to it's initial state, which I think is problematic.

File: lib/puppet_x/msutter/templates/invoke_dsc_resource.erb [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+          Settings
+          {
+              RefreshMode = $mode
+          }
+      }
+  }
+
+  LCMSettings -OutputPath $configOutputPath
+
+  Set-DscLocalConfigurationManager -Path $configOutputPath
```
Comment (Iristyle) [inline lib/puppet_x/msutter/templates/invoke_dsc_resource.erb]: What happens when this call fails (i.e. if there is another operation in progress)?

File: build/dsc/templates/dsc_type_spec.rb.erb [hunk 7]
```diff
@@ -162,26 +162,20 @@ describe Puppet::Type.type(:dsc_<%= resource.friendlyname.downcase %>) do
 
     describe "when dscmeta_import_resource is true (default) and dscmeta_module_name existing/is defined " do
 
-      it "should compute powershell dsc test script with Import-DscResource" do
-        expect(@provider.ps_script_content('test')).to match(/Import-DscResource/)
+      it "should compute powershell dsc test script with Invoke-DscResource" do
+        expect(@provider.ps_invoke_content('test')).to match(/Invoke-DscResource/)
```
Comment (Iristyle) [inline build/dsc/templates/dsc_type_spec.rb.erb]: Ditto changes re: `ps_script_content` vs `ps_invoke_content`

File: lib/puppet/provider/base_dsc/powershell.rb (hunk 2 above)
Comment (jpogran) [inline lib/puppet/provider/base_dsc/powershell.rb]: It wasn't clear to me whether we reached consensus to keep `ps_script_content` or not. If that's the case, then I'll change back to that

File: lib/puppet_x/msutter/helpers/dsc_provider_helpers.rb (hunk 1 above)
Comment (jpogran) [inline lib/puppet_x/msutter/helpers/dsc_provider_helpers.rb]: Nope. Oversight. Removing now.

**Comment by ferventcoder** [comment]
`\o/`

**Pull request closed without merging**